src/App.tsx

```diff
@@ -15,47 +15,114 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  NONE,
+  ALPHABET,
+  LENGTH,
+}
+
+type IsReserved = boolean;
+
 export const App: React.FC = () => {
+  const [isReversed, setReserved] = useState(false);
+  const [sortType, setSortType] = useState(SortType.NONE as SortType);
+
+  const getReorderedGoods = (
+    goods: string[],
+    oldSortType: SortType,
+    oldIsReversed: IsReserved,
+  ) => {
+    const visibleGoods = [...goods];
+
+    visibleGoods.sort((a, b) => {
+      switch (oldSortType) {
+        case SortType.ALPHABET:
+          return a.localeCompare(b);
+        case SortType.LENGTH:
+          return a.length - b.length;
+        default:
+          return SortType.NONE;
+      }
+    });
+
+    if (oldIsReversed) {
+      visibleGoods.reverse();
+    }
+
+    return visibleGoods;
+  };
+
+  const sortAlphabet = () => {
+    setSortType(SortType.ALPHABET);
+  };
+
+  const sortLength = () => {
+    setSortType(SortType.LENGTH);
+  };
+
+  const reverse = () => {
+    setReserved(oldIsReversed => !oldIsReversed);
+  };
+
+  const reset = () => {
+    setSortType(SortType.NONE);
+    setReserved(false);
+  };
+
+  const visibleGoods = getReorderedGoods(
+    goodsFromServer, sortType, isReversed,
+  );
+
   return (
     <div className="section content">
       <div className="buttons">
         <button
           type="button"
-          className="button is-info is-light"
+          className={sortType === SortType.ALPHABET
+            ? 'button is-info'
+            : 'button is-info is-light'}
+          onClick={sortAlphabet}
         >
           Sort alphabetically
         </button>
 
         <button
           type="button"
-          className="button is-success is-light"
+          className={sortType === SortType.LENGTH
+            ? 'button is-success'
+            : 'button is-success is-light'}
+          onClick={sortLength}
         >
           Sort by length
         </button>
 
         <button
           type="button"
-          className="button is-warning is-light"
+          className={isReversed === true
+            ? 'button is-warning'
+            : 'button is-warning is-light'}
+          onClick={reverse}
         >
           Reverse
         </button>
 
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
+        {(isReversed || sortType !== SortType.NONE) && (
```

Comment:
you can store this check statement into variable before return () as e.g.:

`const showResetButton = isReversed || sortType !== SortType.NONE;`

and then use it here to make it cleaner